scripts/periodic_cleanup.py

```diff
@@ -7,62 +7,125 @@
+import libcloud
+
+
+def log(f):
+    @wraps(f)
+    def wrapper(*args, **kwargs):
+        print("Starting: {function}".format(function=f.func_name))
+        result = f(*args, **kwargs)
+        print("Completed: {function}".format(function=f.func_name))
+        return result
+    return wrapper
+
+
+class Cleanup:
+    def __init__(self):
+        self.read_env_vars()
+        self.maas = self.init_rackspace_monitoring()
+        self.jenkins_client = jenkins_node.get_jenkins_client()
+
+    @log
+    def read_env_vars(self):
+        self.age_limit = int(os.environ.get("INSTANCE_AGE_LIMIT", 48))
+        self.instance_prefix = os.environ.get("INSTANCE_PREFIX")
+        if not self.instance_prefix:
+            raise ValueError("INSTANCE_PREFIX must not be empty")
+
+        self.username = os.environ["PUBCLOUD_USERNAME"]
+        self.api_key = os.environ["PUBCLOUD_API_KEY"]
+        self.regions = os.environ["REGIONS"].split(' ')
+
+    @log
+    def init_rackspace_monitoring(self):
+        Cls = get_driver(Provider.RACKSPACE)
+        return Cls(self.username, self.api_key)
+
+    @log
+    def init_pyrax(self, region):
+        pyrax.set_setting("identity_type", "rackspace")
+        pyrax.set_credentials(self.username, self.api_key)
+        self.cs = pyrax.connect_to_cloudservers(
+            region, verify_ssl=True)
+
+    @log
+    def cache_servers(self):
+        self.servers = self.cs.servers.list()
+        self.prefixed_servers = (
+            server for server in self.servers
+            if re.match(self.instance_prefix, server.name)
+        )
+
+    @log
+    def cleanup_instances(self):
+        """ Delete instances if they are in an error state or are over the
+        age defined in INSTANCE_AGE_LIMIT.
+
+        Instances that don't match INSTANCE_PREFIX are ignored.
+        """
+        current_time = datetime.datetime.now(tzutc())
+        max_age = datetime.timedelta(hours=self.age_limit)
+
+        for server in self.prefixed_servers:
+            created_time = dateutil.parser.parse(server.created)
+            age = current_time - created_time
+            errored = server.status == "ERROR"
+            if errored or age > max_age:
+                print("Deleting {name} Errored: {error} Age: {age}".format(
+                    name=server.name, error=errored, age=age))
+                server.delete()
+
+    @log
+    def cleanup_maas_entities(self):
+        """ Remove maas entities that relate to deleted instances """
+        entities = self.maas.list_entities()
+        for entity in entities:
+            # multi node aio entities are vm.hostname eg infra1.omnantp-52-66d1
+            # standard aio entities don't include a vm eg: ramcp-77-5dcd
+            hostname = entity.label.split('.')[-1]
+            # entities with URIs are tied to cloud servers and cannot be
+            # manually deleted
+            if re.match(self.instance_prefix, hostname) and not entity.uri:
```

Comment:
Good catch, It won't delete any entities that are linked to instances, but it will delete indirectly linked entities (eg entities for MNAIO VMs). 

I'll revise this tomorrow.